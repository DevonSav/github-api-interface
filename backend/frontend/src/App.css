.App {
	text-align: center;
}
.App-header {
	background-color: #282c34;
	min-height: 100vh;
	display: flex;
	flex-direction: column;
	align-items: center;
	/*justify-content: center;*/
	font-size: calc(10px + 2vmin);
	color: white;
}
.App-link {
	color: #61dafb;
}


.external_link {
	color: #61dafb;
	text-decoration: none;
}

#user_page_link {
	text-decoration: none;
}

.search_data_input {
	padding: 8px 8px;
	font-size: calc(1px + 1.8vmin);
}

.action_btn {
	padding: 15px 32px;
	margin: 4px 10px;
	color: rgb(240, 235, 226);
	background-color: rgb(31, 33, 33);
	border-color: #585d5d;
	border-style: solid;
	text-align: center;
	text-decoration: none;
	font-size: calc(1px + 1.8vmin);
	font-weight: bold;
	cursor: pointer;
	-webkit-user-select: none;	/* Safari */
	-ms-user-select: none;	/* IE 10 and IE 11 */
	user-select: none;	/* Standard syntax */
	border-radius: 5px;
}
#user_search_btn {
	padding-top: 8px;
	padding-bottom: 8px;
}

#refresh-btn {
	padding-top: 8px;
	padding-bottom: 8px;
}

.user_profile {
	height: auto;
	margin-bottom: 32px;
	border-radius: 5px;
	border: solid whitesmoke;
}
.user_profile .profile_image {
	float: left;
	width: 140px;
	height: 140px;
	margin: 20px;
	border: solid #61dafb;
	border-radius: 50%;
	background: #555;
	cursor: pointer;
}
.user_profile .profile_username {
	position: relative;
	margin: 0;
	top: 30px;
	left: 0px;
	text-align: left;
	font-weight: bold;
}
.user_profile .profile_bio {
	text-align: left;
}

.repo_table .tableHeader {
	padding: 5px 10px;
	padding-left: 0;
	padding-right: 32px;
	padding-top: 5px;
	padding-bottom: 5px;
	text-align: left;
}
.repo_table .repoTableData {
	padding: 5px 10px;
	padding-left: 0;
	padding-right: 32px;
	padding-top: 5px;
	padding-bottom: 5px;
	text-align: left;
}
.repo_table #github_link {
	padding-right: 20px;
}

.user_table {
	border-collapse: collapse;
	border-spacing: 0;
	background-color: #3e4450;
	border-radius: 5px;
}

.user_table .user_table_data a {
	padding-left: 30px;
}
.user_table .tableHeader {
	padding: 5px 10px;
	padding-left: 1px;
	padding-right: 32px;
	padding-top: 5px;
	padding-bottom: 5px;
	text-align: left;
}
.user_table .user_profile_image {
	height: 80px;
	width: 80px;
	margin-right: 20px;
	background: #555;
	border: solid #61fbdf;
	border-radius: 50%;
	cursor: pointer;
}
.user_table #user_page_link {
	padding-left: 0;
	padding-right: 30px;
	font-weight: bold;
	color: #00aeff;
}
