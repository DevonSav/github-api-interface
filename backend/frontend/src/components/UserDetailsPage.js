import React from 'react';
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome';
import { faGithub, faGithubSquare } from '@fortawesome/free-brands-svg-icons';
import { faBars, faSpinner } from '@fortawesome/free-solid-svg-icons';

import logo from '../logo.svg';
import { Link, useParams } from "react-router-dom";
const loadingIcon = <FontAwesomeIcon icon={faSpinner} size='lg' style={{color: "#61dafb",}}/>;


export class UserDetailsPage extends React.Component {
	constructor(props) {
		super(props);
		this.state = {
			error: null,
			isLoaded: false,
			reposList: [],
			commitList: [{message: 'Please Select A Repo'}],
			currentRepoDesc: 'Please Select A Repo',
			userData: {"bio": 'No available user bio.'}
		};
		this.refreshRepoList = this.refreshRepoList.bind(this);
		this.getUserData = this.getUserData.bind(this);
		this.setRepoDescription = this.setRepoDescription.bind(this);
	}

	refreshRepoList() {
		fetch(`/repo-search?name=${this.props.name}`)
			.then(res => res.json())
			.then((response) => {
					this.setState({isLoaded: true});

					this.setState({
						reposList: response.sort((a, b) => a.name - b.name)
					});
				},
				(error) => {
					console.log(error.message);
					this.setState({
						isLoaded: false,
						error
					});
				}
			)
	}

	getUserData() {
		fetch(`/user-search?name=${this.props.name}`)
			.then(res => res.json())
			.then((response) => {
					this.setState({
						userData: response
					});
				},
				(error) => {
					console.log(error.message);
				}
			)
	}


	/**
     * Runs automatically.
     */
    componentDidMount() {
		this.getUserData();
        this.refreshRepoList();
    }

	setRepoDescription(e, repoName) {
		e.preventDefault();

		// While data is being fetched
		this.setState({
			currentRepoDesc: loadingIcon,
			commitList: [{message: loadingIcon}]
		});

		fetch(`/commit-search?name=${this.props.name}&repo=${repoName}`)
			.then(res => res.json())
			.then((response) => {
					this.setState({
						commitList: response
					});
				},
				(error) => {
					console.log('refreshCommitList: ', error.message);
				}
			)

		const repositoryList = this.state.reposList;
		for (let i = 0; i < repositoryList.length; i++) {
			const repo = repositoryList[i];
			if (repo.name == repoName) {
				if (repo.description == null) {
					this.setState({
						currentRepoDesc: (`No description for '${repoName}'`)
					});
				} else {
					this.setState({
						currentRepoDesc: repo.description
					});
				}
			}
		}
		console.log('Set repo description')
	}


	render() {
		const { error, isLoaded, reposList, commitList, currentRepoDesc, userData} = this.state;
		const avatarImage = (userData.avatar_url ? userData.avatar_url : logo);
		const profileBio = (userData.bio ? userData.bio : 'No available user bio.');
		const githubIcon = <FontAwesomeIcon icon={faGithubSquare} size='lg' style={{color: "#61fbdf",}}/>;
		const barIcon = <FontAwesomeIcon icon={faBars} size='lg' style={{color: "#61dafb",}}/>;


		let reposDataList = <h3>ERROR, Something went wrong.</h3>;
		if (error) {
			reposDataList = (<tr><td>Error: {error.message}</td></tr>);
		} else if (!isLoaded) {
			reposDataList = (<tr><td>{loadingIcon} Loading...</td></tr>);
		} else {
			reposDataList = (
				<>
					{reposList.map(repo => (
						<tr key={repo.id} className='repoDataRow'>
							<td className='repo_table_data'><a className='external_link' href={repo.html_url}>{githubIcon} {repo.name}</a></td>
							<td className='repo_table_data'>{repo.updated_at.slice(0, 10)}</td>{/*2023-05-02T13:11:20Z*/}
							<td className='repo_table_data'>{repo.created_at.slice(0, 10)}</td>
							<td className='repo_table_data' id='github_link'><button className='action_btn desc_btn' onClick={(e) => this.setRepoDescription(e, repo.name)}>{barIcon}</button></td>
						</tr>
						))
					}
				</>
			);
		}
		let commitDataList = (
			<>
				{commitList.map(commit => (
						<li key={commit.sha} className='list_data'>{commit.message}</li>
					))
				}
			</>
		);

		let repoDisplaySection = (
			<div className='repo_list_group'>
				<h2 id='no_available_repos'>No publicly available repositories.</h2>
			</div>
		);
		if (reposList.length > 0) {
			repoDisplaySection = (
				<>
				<button className='action_btn repo_search' onClick={(e) => this.refreshRepoList(e)}>Refresh Repo List</button>
				<div className='repo_list_group'>
					<table className='repo_table'>
						<thead>
							<tr>
								<th className='table_header'>Repository</th>
								<th className='table_header'>Last commit</th>
								<th className='table_header'>Created</th>
								<th className='table_header' id='repo_desc_btn'></th>
							</tr>
						</thead>
						<tbody>
							{reposDataList}
						</tbody>
					</table>
					<div className='repo_detail_div'>
						<div className='repo_description'>
							<h3 className='repo_detail_header'>Description</h3>
							<p className='repo_contents'>{currentRepoDesc}</p>
						</div>
						<div className='repo_commits_div'>
							<h3 className='repo_detail_header'>Recent commits</h3>
							<ul className='commit_list'>
								{commitDataList}
							</ul>
						</div>
					</div>
				</div>
				</>
			);
		}


		return (
			<div className="github_interface">
				<Link to='/' className="action_btn home_btn">BACK TO SEARCH</Link>
				<h1 id="main_heading">User Details</h1>
				<div className='user_profile'>
				<a href={`https://github.com/${this.props.name}`}><img className="profile_image" alt="profile" src={avatarImage}/></a>
					<h3 className='profile_username'>{this.props.name}</h3>
					<br></br>
					<p className='profile_bio'>{profileBio}</p>
				</div>

				{repoDisplaySection}

			</div>
		);
	}
}


/**
 * Displays user details.
 * @param {*} params name
 */
const UserDetailsComponent = () => {
	const {loginName} = useParams();
	const profileDiv = (
		<UserDetailsPage name={loginName}/>
	);
	return profileDiv;
}
export default UserDetailsComponent;

/** REFERENCE
 * https://stackoverflow.com/questions/30115324/pass-props-in-link-react-router
 */
