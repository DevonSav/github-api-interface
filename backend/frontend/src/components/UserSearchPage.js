import React from 'react';

export default class UserSearchPage extends React.Component {
	constructor(props) {
		super(props);
		this.state = {
			error: null,
			isLoaded: false,
			usersList: [],
			nameInput: null
		};
		this.refreshUsersList = this.refreshUsersList.bind(this);
		this.handleChange = this.handleChange.bind(this);
	}


	refreshUsersList(event) {
		event.preventDefault(); // prevent page refresh
		let newShortUserList = [];

		fetch(`/user-search?name=${this.state.nameInput}`)
			.then(res => res.json())
			.then((response) => {
					console.log(response);
					newShortUserList.push(response);
					let hasName = response.name ? true : false;
					this.setState({isLoaded: true});

					if (hasName) {
						this.setState({
							usersList: newShortUserList.sort((a, b) => a.name - b.name)
						});
					} else {
						this.setState({
							usersList: newShortUserList.sort((a, b) => a.login - b.login)
						});
					}
				},
				(error) => {
					console.log(error.message);
					this.setState({
						isLoaded: false,
						error
					});
				}
			)
	}

	/**
	 * Gets the input value and sets the corresponding state variable.
	 */
	handleChange(event) {
		const value = event.target.value;
		const inputFieldElementId = event.target.id;
		this.setState({
			[inputFieldElementId]: value
		});
	}

	render() {
		let usersDataList = <h3>ERROR, Something went wrong.</h3>;
		const { error, isLoaded, usersList} = this.state;

		if (error) {
			usersDataList = (<tr><td>Error: {error.message}</td></tr>);
		} else if (!isLoaded) {
			usersDataList = (<tr><td>Waiting for search...</td></tr>);
		} else {
			//If the name is null use the login value
			usersDataList = (
				<>
					{usersList.map(user => (
						<tr key={user.id} className='userDataRow'>
							<td className='userTableData'><img className="user-profile-image" alt="profile" src={user.avatar_url}/></td>
							<td className='userTableData'>{user.name ? user.name : user.login}</td>
						</tr>
						))
					}
				</>
			);
		}

		return (
			<div className="github-interface">
				<h2 id="main-heading">Github Interface</h2>
				<form>
					<input className='search-data-input' id='nameInput' type="text" placeholder='Name' onChange={e => this.handleChange(e)}/>
					<button id='user-search-btn' className='action-btn' onClick={(e) => this.refreshUsersList(e)}>Search</button>
				</form>

				<br></br>
				<table className='user-table'>
					<thead>
						<tr>
							<th className='tableHeader'></th>
							<th className='tableHeader'>Name</th>
						</tr>
					</thead>
					<tbody>
						{usersDataList}
					</tbody>
				</table>
			</div>
		);
	}
}

/** REFERENCES
 *
 */