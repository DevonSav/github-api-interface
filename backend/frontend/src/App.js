import './App.css';
import UserSearchPage from './components/UserSearchPage';
import UserDetailsPage from './components/UserDetailsPage';
import {BrowserRouter, Routes, Route, Link} from 'react-router-dom';

function App() {
    return (
        <div className="App">
            <header className="App-header">
            <BrowserRouter>
                <Routes>
                    <Route exact path="/" element={<UserSearchPage />} />
                    <Route path="/user-details/:loginName" element={<UserDetailsPage />} />
                </Routes>
            </BrowserRouter>
            </header>
        </div>
    );
}

export default App;
