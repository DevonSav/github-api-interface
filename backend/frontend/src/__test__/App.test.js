//import { render, screen } from '@testing-library/react';
import App from '../App';
import renderer from 'react-test-renderer';
import { BrowserRouter } from "react-router-dom";
import UserSearchPage from '../components/UserSearchPage';
import UserDetailsPage from '../components/UserDetailsPage';

test('renders correctly', () => {
    const tree = renderer
    .create(<App />)
    .toJSON();
    expect(tree).toMatchSnapshot();
});


//test('renders learn react link', () => {
//  render(<App />);
//  const linkElement = screen.getByText(/learn react/i);
//  expect(linkElement).toBeInTheDocument();
//});
