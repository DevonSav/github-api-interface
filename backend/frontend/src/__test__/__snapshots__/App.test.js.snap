// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<div
  className="App"
>
  <header
    className="App-header"
  >
    <div
      className="github_interface"
    >
      <h2
        id="main_heading"
      >
        Github Interface
      </h2>
      <form
        className="user_search_form"
      >
        <input
          className="search_data_input"
          id="nameInput"
          onChange={[Function]}
          placeholder="Name"
          type="text"
        />
        <button
          className="action_btn user_search"
          onClick={[Function]}
        >
          Search
        </button>
      </form>
      <br />
      <table
        className="user_table"
      >
        <thead>
          <tr>
            <th
              className="table_header left"
            >
              Github
            </th>
            <th
              className="table_header"
            >
              Name
            </th>
          </tr>
        </thead>
        <tbody>
          <tr>
            <td />
            <td
              id="user_search_waiting"
            >
              Waiting for search...
            </td>
          </tr>
        </tbody>
      </table>
    </div>
  </header>
</div>
`;
